### PR TITLE
Display simple coverage report to console

### PR DESCRIPTION
**Pull request by hauleth**
This should reduce need for packages like excoveralls in some use cases
when all what is needed is to have total coverage percentage in any
form available for CI to provide coverage badge.

Reference: https://docs.gitlab.com/ee/user/project/pipelines/settings.html#test-coverage-parsing

### CODE REVIEW

**Comment by josevalim** [comment]
Can you please include an example of the report? Does any other client besides gitlab parses coverage reports like that?

File: lib/mix/lib/mix/tasks/test.ex [hunk 1]
```diff
@@ -2,6 +2,8 @@ defmodule Mix.Tasks.Test do
   defmodule Cover do
     @moduledoc false
 
+    @threshold 90
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: This would need to be made configurable.

File: lib/mix/lib/mix/tasks/test.ex [hunk 2]
```diff
@@ -17,14 +19,64 @@ defmodule Mix.Tasks.Test do
...
+    defp display({name, coverage}, opts) do
+      threshold = Keyword.get(opts, :threshold, @threshold)
+      percentage = percentage(coverage)
+
+      Mix.shell().info([
+        colour(percentage, threshold),
+        format(percentage, 9),
+        "%",
+        :reset,
+        " | #{name}"
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: You need to inspect the module name so it shows only as "MyApp.Module" and not "Elixir.MyApp.Module". But you need to be careful to not inspect the string "Total".

**Comment by josevalim** [comment]
Also please check the build.

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: This is configurable via `threshold` value in `test_coverage` configuration option. I haven't documented it yet thought.

**Comment by josevalim** [comment]
Let's also provide an option that skips the summary if desired.

File: lib/mix/lib/mix/tasks/test.ex (hunk 2 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: That is the reason why I haven't done it, however I can make it happen.

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (josevalim) [reply lib/mix/lib/mix/tasks/test.ex]: Beautiful. Maybe we should call this `@default_threshold` then?

**Comment by josevalim** [comment]
One last thing: gitlab default's regex expects the word "covered" in there. Do we need to add something similar? Wouldn't just printing percentages be a bit ambiguous? What regex are you using currently?

**Comment by hauleth** [comment]
@josevalim about comments, from top to bottom:

- I am not aware of any that works in such way, however I only skimmed through available options, so maybe there are ones that I am no aware of.
- I need to update tests.
- I was thinking about something other: sort solutions based on their coverage percentage and then display only "bottom N" of options. Then such option could be set to `false` to disable CLI coverage report at all. What do you think? Or rather add additional option that would disable output and separate that would print sorted results?
- Default regex yes, but this can be parsed using `\d+(\.\d+)?%\s*\|\s*Total`.

**Comment by hauleth** [comment]
Example result:

```
Cover results
Percentage | Module
-----------|--------------------------
      0.0% | Elixir.Cover
      0.0% | Elixir.ImagerWeb.Instrumenter
     50.0% | Elixir.Imager.Application
      0.0% | Elixir.Imager.Convert
      0.0% | Elixir.ImagerWeb.ChannelCase
      0.0% | Elixir.ImagerWeb.UserSocket
      0.0% | Elixir.ImagerWeb.WebhookController
     6.45% | Elixir.ImagerWeb.Router.Helpers
    15.79% | Elixir.ImagerWeb.ErrorView
    45.45% | Elixir.ImagerWeb.ConnCase
     5.26% | Elixir.ImagerWeb.Endpoint
    23.81% | Elixir.Imager.Stats
     5.71% | Elixir.Imager
      0.0% | Elixir.ImagerWeb.Router
      0.0% | Elixir.ImagerWeb
-----------|--------------------------
     7.88% | Total
```

**Comment by josevalim** [comment]
> I was thinking about something other: sort solutions based on their coverage percentage and then display only "bottom N" of options. Then such option could be set to false to disable CLI coverage report at all.

To me it is OK to sort by percentage but is there a reason to allow a cut off?

**Comment by michalmuskala** [comment]
Should we print the module names with elixir syntax?

**Comment by hauleth** [comment]
@michalmuskala @josevalim how about such output:

```
Coverage results:

Percentage | Module
-----------|--------------------------
      0.0% | Imager.Convert
      0.0% | Cover
      0.0% | ImagerWeb.Instrumenter
      0.0% | ImagerWeb
      0.0% | ImagerWeb.ChannelCase
      0.0% | ImagerWeb.UserSocket
      0.0% | ImagerWeb.WebhookController
      0.0% | ImagerWeb.Router
     5.26% | ImagerWeb.Endpoint
     5.71% | Imager
     6.45% | ImagerWeb.Router.Helpers
    15.79% | ImagerWeb.ErrorView
    23.81% | Imager.Stats
    45.45% | ImagerWeb.ConnCase
     50.0% | Imager.Application
-----------|--------------------------
     7.73% | Total

Generate coverage results in cover
```

**Comment by hauleth** [comment]
I am thinking about changing the last line to `Generated HTML coverage results in 'cover' directory`.

File: lib/mix/lib/mix/tasks/test.ex (hunk 1 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: No problem

**Comment by webofbits** [comment]
@hauleth could you show us what an example of the HTML page would look like?

**Comment by hauleth** [comment]
@webdeb HTML result is left as is. For more information check Erlang's `cover` module.

**Comment by webofbits** [comment]
You coverage logs are fine. Not sure if would be possible to have some additional info like lines, relevant or missed but I think that would be great to have. Just an opinion though and thanks for this pull request. 
```
COV    FILE                                        LINES RELEVANT   MISSED
100.0% lib/module_name.ex                              9        0        0
```

File: lib/mix/lib/mix/tasks/test.ex [hunk 3]
```diff
@@ -17,14 +19,64 @@ defmodule Mix.Tasks.Test do
       output = opts[:output]
 
       fn ->
-        Mix.shell().info("\nGenerating cover results ...")
         File.mkdir_p!(output)
 
-        Enum.each(:cover.modules(), fn mod ->
-          {:ok, _} = :cover.analyse_to_file(mod, '#{output}/#{mod}.html', [:html])
-        end)
+        Mix.shell().info("\nCover results")
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: We can keep this line as it was before because it will also work if folks choose to skip the summary.

**Comment by hauleth** [comment]
Yes this is perfectly possible, however I want to keep output format as simple as possible without any additional clutter that can be provided by external tools.

@josevalim @michalmuskala any opinions about output format?

**Comment by josevalim** [comment]
@hauleth I like your suggested output last. I would just keep the first line of the output and the last line the same as it is today. Let's minimiza changes and make sure the summary is optional.

@webofbits let's start simple for now we can always add more information later.

File: lib/mix/lib/mix/tasks/test.ex (hunk 3 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: How about keeping this line and adding additional output after HTML result was generated?

**Comment by hauleth** [comment]
@josevalim @webofbits 

```
Generating cover results ...

Coverage results:

Percentage | Module
-----------|--------------------------
      0.0% | Imager.Convert
      0.0% | ImagerWeb.ChannelCase
      0.0% | ImagerWeb.UserSocket
      0.0% | ImagerWeb
      0.0% | ImagerWeb.WebhookController
      0.0% | Cover
      0.0% | ImagerWeb.Instrumenter
      0.0% | ImagerWeb.Router
     5.26% | ImagerWeb.Endpoint
     5.71% | Imager
     6.45% | ImagerWeb.Router.Helpers
    15.79% | ImagerWeb.ErrorView
    23.81% | Imager.Stats
    45.45% | ImagerWeb.ConnCase
     50.0% | Imager.Application
-----------|--------------------------
     7.65% | Total

Generated HTML coverage results in 'cover' directory
```

**Comment by webofbits** [comment]
Sounds good to me :+1:

**Comment by josevalim** [comment]
@hauleth "Coverage results:" is unnecessary. From the previous line and context we know they are results.

File: lib/mix/lib/mix/tasks/test.ex [hunk 4]
```diff
@@ -1,6 +1,22 @@
 defmodule Mix.Tasks.Test do
   defmodule Cover do
-    @moduledoc false
+    @default_formatters [:console, :html]
+    @default_threshold 90
+
+    @moduledoc """
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: Let's please keep this module still private. I don't think we should expose it. Let's just have a section in the `mix test` docs that says the configurations for the default cover handling.

File: lib/mix/lib/mix/tasks/test.ex [hunk 5]
```diff
@@ -14,17 +30,91 @@ defmodule Mix.Tasks.Test do
           Mix.raise("Failed to cover compile directory: " <> compile_path)
       end
 
-      output = opts[:output]
+      formatters = Keyword.get(opts, :formatters, @default_formatters)
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: `:console` is not really a formatter because it doesn't print it for every module. It is just a summary. So I would rather have a `summary: true | false` option.

File: lib/mix/lib/mix/tasks/test.ex [hunk 6]
```diff
@@ -14,17 +30,91 @@ defmodule Mix.Tasks.Test do
...
 
       fn ->
+        {:result, ok, _fail} = :cover.analyse(:coverage, :module)
         Mix.shell().info("\nGenerating cover results ...")
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/test.ex]: Let's print this before we call analyze, otherwise the user may wait for something and they don't know what it is.

File: lib/mix/lib/mix/tasks/test.ex (hunk 5 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: According to Erlang `cover:analyse/2` docs it will print data for all cover compiled modules. However I can misunderstanding your comment.

Comment (josevalim) [reply lib/mix/lib/mix/tasks/test.ex]: Do you mean that we have also added an option to print all of the results to the terminal? If so, that's a different feature than the one we were debating. Let's focus on the summary feature first. Other additions should be considered separately.

Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: That is the exact result of what this PR produces https://github.com/elixir-lang/elixir/pull/7664#issuecomment-387730487. It display all modules and their respective coverage percentage. You meant something else? Like we should display only overall coverage like:

```
Generating cover results ...

Coverage: 72.32%

Generated HTML coverage results in 'cover' directory
```

?

Comment (josevalim) [reply lib/mix/lib/mix/tasks/test.ex]: Ok. Sorry for the misunderstanding. The current output is fine. However, I wouldn't call it a formatter because what we print to the console and what we write to html files are two very distinct things. My proposal is to remove the idea of formatters and always generate HTML as before. Let's call the new reporting that you have added as part of this PR "Summary" and it can be disabled via the `summary: true | false` option.

**Comment by whatyouhide** [comment]
I might be missing something obvious, but why have the percentage first (left) and the module name after (right) instead of the other way around?

**Comment by OvermindDL1** [comment]
@whatyouhide The percentage is pretty bounded in how many characters it can take up, where a module name is not (in comparison), thus I think it is easier to keep both readable instead of the module name occasionally pushing the percentages off the side with long module names.  :-)

**Comment by hauleth** [comment]
Exactly what @OvermindDL1 said. It is easier to display it in that way instead of looping through module names twice and building list based on the longest module name.

**Comment by whatyouhide** [comment]
Yeah I get that it's easier to build but is that an argument for the user of this or for the implementer? I think the difference in implementation complexity is so small that it justifies having what IMO feels more natural to the user, that is, modules first and percentage after. Navigating the list of modules twice shouldn't be a performance issue since these lists are supposed to be on the "very small" side.

**Comment by josevalim** [comment]
I think the most important information here is the percentage and given that we will be ordering by %, I vote to keep it first too. The goal here is not to list modules.

**Comment by OvermindDL1** [comment]
As the user I find the percentages first *significantly* more preferable as well as I want to see what they are before I care about module names.

**Comment by hauleth** [comment]
@whatyouhide I also find it much easier to connect module with percentage in that format as these are next to each other, if there would be bunch of spaces between them and there would be some long and short module names then it would be hard to follow lines without some kind of guide. While it is fairly easy to get guide in your terminal view then it is pretty hard to get one in CI output if the CI UI author didn't thought about it beforehand.

**Comment by hauleth** [comment]
Using example from above:

```
 Module                      | Percentage
-----------------------------|------------
 Imager.Convert              |       0.0%
 ImagerWeb.ChannelCase       |       0.0%
 ImagerWeb.UserSocket        |       0.0%
 ImagerWeb                   |       0.0%
 ImagerWeb.WebhookController |       0.0%
 Cover                       |       0.0%
 ImagerWeb.Instrumenter      |       0.0%
 ImagerWeb.Router            |       0.0%
 ImagerWeb.Endpoint          |      5.26%
 Imager                      |      5.71%
 ImagerWeb.Router.Helpers    |      6.45%
 ImagerWeb.ErrorView         |     15.79%
 Imager.Stats                |     23.81%
 ImagerWeb.ConnCase          |     45.45%
 Imager.Application          |      50.0%
-----------------------------|------------
 Total                       |      7.65%
```

It can get pretty hard to connect `Imager` with `5.71%`, and now imagine if there would be block of short module names and there would be one exceptionally long. It could became very hard to connect values with names on the first sight.

**Comment by hauleth** [comment]
Only open question is whether sort in increasing or decreasing order. I do not think it should be configureable, but there should be defined order.

**Comment by hauleth** [comment]
@whatyouhide another reason why it is better to leave percentage in front is that it is much easier to crop lines at the end in the terminal window rather than crop them in the middle or at the beginning to have visible results in case of very long module names.

**Comment by OvermindDL1** [comment]
> Only open question is whether sort in increasing or decreasing order. I do not think it should be configureable, but there should be defined order.

Honestly I'd prefer decreasing so that the 0%'ers are at the bottom as that is the most immediate thing I will see when running commands on the terminal.

**Comment by AndrewDryga** [comment]
@OvermindDL1 I think it's much better to see them in increasing (ascending) order. I feel like it's better to have issues to be on top of the list (first things should be addressed first), not on the bottom of it.

**Comment by OvermindDL1** [comment]
@AndrewDryga Except with them being at the top then the issues have scrolled off the screen and I can't see them without doing `<Ctrl>+<B>, <[>` and then scroll around with the keys, that's a pain to see the information that I want to see most, where the last printed information is the information that is immediately viewable and accessible.

File: lib/mix/lib/mix/tasks/test.ex [hunk 7]
```diff
@@ -162,6 +232,14 @@ defmodule Mix.Tasks.Test do
   It must return either `nil` or an anonymous function of zero arity that will
   be run after the test suite is done.
 
+  ### Default coverage module options
+
+    * `:threshold` - summary output colours resulting percentage red/green
+      depending on weather given value is below or over given threshold,
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: s/weather/whether/

File: lib/mix/lib/mix/tasks/test.ex [hunk 8]
```diff
@@ -162,6 +232,14 @@ defmodule Mix.Tasks.Test do
   It must return either `nil` or an anonymous function of zero arity that will
   be run after the test suite is done.
 
+  ### Default coverage module options
+
+    * `:threshold` - summary output colours resulting percentage red/green
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: The `:threshold` option seems to only make sense when the `:summary` option is enabled. In order to reflect that hierarchy, what do you think about?

```elixir
test_coverage: [summary: true] # Enables the summary with the default threshold
test_coverage: [summary: [threshold: 70]] # Enables the summary and sets the threshold to 70
```

If in the future the summary adds more configuration variables, they won't be mixed with the global coverage options.

File: lib/mix/lib/mix/tasks/test.ex [hunk 9]
```diff
@@ -162,6 +232,14 @@ defmodule Mix.Tasks.Test do
   It must return either `nil` or an anonymous function of zero arity that will
   be run after the test suite is done.
 
+  ### Default coverage module options
```
Comment (fertapric) [inline lib/mix/lib/mix/tasks/test.ex]: Should these options be listed in the "Coverage" section above?

> ## Coverage
>
> The `:test_coverage` configuration accepts the following options:
>
>  * `:output` - the output for cover results, defaults to `"cover"`
>  * `:tool`   - the coverage tool
>  * `:summary` - whether reporter should output coverage data summary to the console, defaults to `true`

**Comment by josevalim** [comment]
Let's go with decreasing order (100%) comes first. Let's also adopt @fertapric's  suggestion.

**Comment by hauleth** [comment]
Everything is updated to what @fertapric requested.

**Comment by lexmag** [comment]
Does it make sense to drop `%` sign given that we explicitly state it is "Percentage"?

File: lib/mix/lib/mix/tasks/test.ex [hunk 10]
```diff
@@ -149,8 +225,14 @@ defmodule Mix.Tasks.Test do
 
   The `:test_coverage` configuration accepts the following options:
 
-    * `:output` - the output for cover results, defaults to `"cover"`
-    * `:tool`   - the coverage tool
+    * `:output`  - the output for cover results, defaults to `"cover"`
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/test.ex]: We don't align `-` anymore due to noise it generates in commits.

File: lib/mix/lib/mix/tasks/test.ex [hunk 11]
```diff
@@ -14,17 +16,91 @@ defmodule Mix.Tasks.Test do
...
       end
+
+      Mix.shell().info([
+        "Generated HTML coverage results in '",
+        output,
+        "' directory\n"
+      ])
     end
+
+    defp colour(percentage, true), do: colour(percentage, @default_threshold)
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/test.ex]: Except "behaviour" we should use American spelling.

File: lib/mix/lib/mix/tasks/test.ex [hunk 12]
```diff
@@ -14,17 +16,91 @@ defmodule Mix.Tasks.Test do
...
+          {tot_cov + cov, tot_not_cov + not_cov}
         end)
+
+      Mix.shell().info("-----------|--------------------------")
+
+      display({"Total", total}, opts)
+      Mix.shell().info("")
+    end
+
+    defp console(_, _), do: nil
```
Comment (lexmag) [inline lib/mix/lib/mix/tasks/test.ex]: Maybe match on `[false, nil]` to make `summary: 123` fail?

**Comment by lexmag** [comment]
@hauleth in my opinion this one character does not improve readability (rather brings visual noise).
I think if we want to add other metrics in the future we'll reconsider the current output quite a bit anyway.

**Comment by lexmag** [comment]
@hauleth just checked the common agreement on this with the rest of the folks in the team: let's keep the sign. 👍

File: lib/mix/lib/mix/tasks/test.ex [hunk 13]
```diff
@@ -14,17 +16,91 @@ defmodule Mix.Tasks.Test do
...
+      |> Integer.to_string()
+      |> String.pad_leading(len)
+    end
+
+    defp format(num, len) when is_float(num) do
+      num
+      |> Float.round(2)
+      |> Float.to_string()
+      |> String.pad_leading(len)
+    end
```
Comment (michalmuskala) [inline lib/mix/lib/mix/tasks/test.ex]: Could we use `:io_lib.format("~#{len}.2f", [num])` for formatting here? This will always give us 2 fields of precision and nicely align the results.

File: lib/mix/lib/mix/tasks/test.ex [hunk 14]
```diff
@@ -14,17 +16,84 @@ defmodule Mix.Tasks.Test do
...
+        :reset,
+        " | ",
+        format_name(name)
+      ])
     end
+
+    defp percentage({0, 0}), do: 100
+    defp percentage({cov, not_cov}), do: cov / (cov + not_cov) * 100
+
+    defp format(num, len) when is_integer(num),
```
Comment (eksperimental) [inline lib/mix/lib/mix/tasks/test.ex]: Generally we use full name in arguments, so it reads better `number` and `length`

File: lib/mix/lib/mix/tasks/test.ex [hunk 15]
```diff
@@ -14,17 +16,84 @@ defmodule Mix.Tasks.Test do
...
+        format(percentage, 9),
+        "%",
+        :reset,
+        " | ",
+        format_name(name)
+      ])
     end
+
+    defp percentage({0, 0}), do: 100
+    defp percentage({cov, not_cov}), do: cov / (cov + not_cov) * 100
```
Comment (eksperimental) [inline lib/mix/lib/mix/tasks/test.ex]: It reads better full name arg. `covered` and `not_covered`

File: lib/mix/lib/mix/tasks/test.ex (hunk 14 above)
Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: @eksperimental my problem with `length` is that vim-elixir colours it as a keyword which I find distracting. Also I find `len` and `num` popular enough that I didn't even thought about using full name.

Comment (OvermindDL1) [reply lib/mix/lib/mix/tasks/test.ex]: > my problem with length is that vim-elixir colours it as a keyword which I find distracting

Wait what why?  It's not a keyword!  o.O

Comment (hauleth) [reply lib/mix/lib/mix/tasks/test.ex]: elixir-editors/vim-elixir#378

File: lib/mix/lib/mix/tasks/test.ex [hunk 16]
```diff
@@ -14,17 +16,85 @@ defmodule Mix.Tasks.Test do
...
+    defp console(results, opts) when is_list(opts) do
+      Mix.shell().info("Percentage | Module")
+      Mix.shell().info("-----------|--------------------------")
+
+      total =
+        results
+        |> Stream.each(&display(&1, opts))
+        |> Enum.reduce({0, 0}, fn {_, {covered, not_covered}},
+                                  {total_covered, total_not_covered} ->
+          {total_cov + cov, total_not_cov + not_cov}
```
Comment (michalmuskala) [inline lib/mix/lib/mix/tasks/test.ex]: This line should be:
```
{total_covered + covered, total_not_covered + not_covered}
```

**Comment by josevalim** [comment]
I have pushed a commit that fixes the build, so as soon as it passes we are good to go!

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
